tests/functional/aws-node-sdk/test/object/corsHeaders.js

```diff
@@ -229,86 +262,126 @@ function _waitForAWS(callback, err) {
-                assert.deepEqual(resHeaders[key], expectedHeaders[key],
-                  `error header: ${key}`);
+                assert.deepEqual(resHeaders[key], expectedHeaders[key], `error header: ${key}`);
             });
         } else {
-            // if no headersResponse provided, should not have these headers
-            // in the request
-            ['access-control-allow-origin',
-                'access-control-allow-methods',
-                'access-control-allow-credentials',
-                'vary'].forEach(key => {
-                    assert.strictEqual(resHeaders[key], undefined,
-                    `Error: ${key} should not have value`);
-                });
+            // if no expectedHeaders provided, should not have these headers in the response
+            ['access-control-allow-origin', 
+            'access-control-allow-methods', 
+            'access-control-allow-credentials', 
+            'vary'].forEach(key => {
+                assert.strictEqual(resHeaders[key], undefined, `Error: ${key} should not have value`);
+            });
         }
-        cb();
     }
-    const method = action.bind(s3);
-    const request = method(params);
-    // modify underlying http request object created by aws sdk to add
-    // origin header
-    request.on('build', () => {
-        request.httpRequest.headers.origin = origin;
-    });
-    request.on('success', response => {
-        const resHeaders = response.httpResponse.headers;
-        _runAssertions(resHeaders, () => {
-            if (response.data.UploadId) {
-                // abort multipart upload before deleting bucket in afterEach
-                return s3.abortMultipartUpload({ Bucket: bucket, Key: objectKey,
-                    UploadId: response.data.UploadId }, callback);
+
+    // Create a new S3 client for each request to avoid middleware conflicts
+    const testS3 = new S3Client(config);
+    let capturedHeaders = {};
+
+    // Add middleware to capture response headers (similar to AWS SDK v2's event approach)
+    testS3.middlewareStack.add(
+        next => async args => {
+            if (origin) {
+                if (!args.request.headers) {
+                    // eslint-disable-next-line no-param-reassign
+                    args.request.headers = {};
+                }
+                // eslint-disable-next-line no-param-reassign
+                args.request.headers['origin'] = origin;
             }
-            return callback();
-        });
-    });
-    // CORS headers should still be sent in case of errors as long as
-    // request matches CORS configuration
-    request.on('error', () => {
-        const resHeaders = request.response.httpResponse.headers;
-        _runAssertions(resHeaders, callback);
-    });
-    request.send();
+
+            try {
+                const result = await next(args);
+
+                // Capture response headers (equivalent to request.on('success'))
```

Comment:
I'm not sure it's really useful here. You have that in mind cause you did v2 to v3, but someone new will not understand